components/camel-kudu/src/main/java/org/apache/camel/component/kudu/KuduProducer.java

```diff
@@ -0,0 +1,114 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.component.kudu;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import org.apache.camel.Exchange;
+import org.apache.camel.support.DefaultProducer;
+import org.apache.kudu.ColumnSchema;
+import org.apache.kudu.Schema;
+import org.apache.kudu.client.CreateTableOptions;
+import org.apache.kudu.client.Insert;
+import org.apache.kudu.client.KuduClient;
+import org.apache.kudu.client.KuduException;
+import org.apache.kudu.client.KuduScanner;
+import org.apache.kudu.client.KuduTable;
+import org.apache.kudu.client.PartialRow;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The Kudu producer.
+ *
+ * @see org.apache.camel.component.kudu.KuduEndpoint
+ */
+public class KuduProducer extends DefaultProducer {
+
+    private static final Logger LOG = LoggerFactory.getLogger(KuduProducer.class);
+
+    private final KuduEndpoint endpoint;
+
+    public KuduProducer(KuduEndpoint endpoint) {
+        super(endpoint);
+
+        if (endpoint == null || endpoint.getKuduClient() == null) {
+            throw new IllegalArgumentException("Can't create a producer when the database connection is null");
+        }
+        this.endpoint = endpoint;
+    }
+
+    @Override
+    public void process(Exchange exchange) throws Exception {
+        String table = endpoint.getTableName();
+        switch (endpoint.getOperation()) {
+        case INSERT:
+            doInsert(exchange, table);
+            break;
+        case CREATE_TABLE:
+            doCreateTable(exchange, table);
+            break;
+        case SCAN:
+            doScan(exchange, table);
+            break;
+        default:
+            throw new IllegalArgumentException("The operation " + endpoint.getOperation() + " is not supported");
+        }
+    }
+
+    private void doInsert(Exchange exchange, String tableName) throws KuduException {
+        KuduClient connection = endpoint.getKuduClient();
+        KuduTable table = connection.openTable(tableName);
+
+        Insert insert = table.newInsert();
+        PartialRow row = insert.getRow();
+
+        Map<?, ?> rows = exchange.getIn().getBody(Map.class);
+        for (Map.Entry<?, ?> entry : rows.entrySet()) {
+            row.addObject(entry.getKey().toString(), entry.getValue());
+        }
+
+        connection.newSession().apply(insert);
+    }
+
+    private void doCreateTable(Exchange exchange, String tableName) throws KuduException {
+        KuduClient connection = endpoint.getKuduClient();
+        LOG.debug("Creating table {}", tableName);
+
+        Schema schema = (Schema) exchange.getIn().getHeader("Schema");
```

Comment:
Here we are expecting a header `Schema`, however I couldn't find some information how do we set it up, do we expect the user to add it?

Reply:
Would it be better to ask for simple text headers that build the schema?

Reply:
What I meant just to indicate in the docs that in order to use Kudu producer to create a table, a schema should be set in the headers, for example [camel-kafka](https://github.com/apache/camel/blob/master/components/camel-kafka/src/main/docs/kafka-component.adoc#producer-headers) and [camel-kniesis](https://github.com/apache/camel/blob/master/components/camel-aws-kinesis/src/main/docs/aws-kinesis-component.adoc#message-headers-used-by-the-kinesis-producer-to-write-to-kinesis--the-producer-expects-that-the-message-body-is-a-byte) docs, so the user can be aware of it. Which format would be better, I honestly have no much of details into kudu, hence I can't tell which is better. What I can recommend is, you can try to create a simple example for a camel route with kudu and see how the user can use it, at least you will get a feeling what can be more convenient for the user.

Reply:
I agree with @omarsmak , you can choose any header name style which is not intersect with another components to avoid conflicts and will be clear to understand for you and other component users. CamelKuduSchema is also suitable or you can define this name yourself.